rfcs/20231017-collective-broadcast.md

```diff
@@ -0,0 +1,109 @@
+
+## Why call it collective_broadcast and not just broadcast?
+Unfortunately, the op name `broadcast` is already taken by [an op in XLA proper](https://www.tensorflow.org/xla/operation_semantics#broadcast), so we can't have the two names clash. `collective_broadcast` was the preferred alternative.
+
+## Proposed Specification
+
+### collective_broadcast
+
+#### Semantics
+
+Within each process group in the StableHLO process grid, send the value of the
+`operand` tensor from the source process to the target processes and produce a
+`result` tensor. For any process `n` not in any replica group, that process will just
+return zeros.
+
+The operation splits the StableHLO process grid into `process_groups` which is
+defined as follows:
+
+* `cross_replica(replica_groups)` if `channel_id <= 0`.
+* `cross_partition(replica_groups)` if `channel_id > 0`.
+
+Afterwards, `result@process` is given by:
+
+* `operand@process_groups[i, 0]` if there exists an `i` such that
+  the process is in `process_groups[i]`.
+* `broadcast_in_dim(constant(0, element_type(result)), [], type(result))`
+  if process is never found in `process_groups`.
+
+
+#### Inputs
+
+| Label | Name                    | Type                                                             | Constraints |
+|-------|-------------------------|------------------------------------------------------------------|-------------|
+| (I1)  | `operand`               | tensor                                                           | (C3)        |
+| (I2)  | `replica_groups`        | variadic number of 1-dimensional tensor constants of type `si64` | (C1-C2)     |
+| (I3)  | `channel_id`            | constant of type `si64`                                          |             |
+
+#### Outputs
+
+| Name     | Type   | Constraints |
+|----------|--------|-------------|
+| `result` | tensor | (C3)        |
+
+#### Constraints
+
+* (C1) is_unique(replica_groups).
+* (C2) `0 <= replica_groups < N` where `N` is defined as:
+  * `num_replicas` if `cross_replica` is used.
+  * `num_partitions` if `cross_partition` is used.
+* (C3) `type(result) = type(operand)`.
+
+#### Examples
```

Comment:
Do you have an example with broadcast having a handle < 0?

Reply:
Good point. Actually, the convention in the spec is to have example with `handle = 0`.

@chaserileyroberts Can you update the spec with this example instead and delete the other two?

We are trying to keep only one example in the spec (see [reference_checklist.md](https://github.com/openxla/stablehlo/blob/main/docs/reference_checklist.md)) and additional ones linked to a test file that the interpreter can run, so these tests will be used some day as long as we have the paper trail for now. I know this type of feedback is more appropriate for the PR, but it would be nice to have the PR merged with no additional feedback.

Proposed test:
```
// num_replicas: 4
// num_partitions: 1
// %operand@(0, 0): [[1, 2]]
// %operand@(1, 0): [[3, 4]]
// %operand@(2, 0): [[5, 6]]
// %operand@(3, 0): [[7, 8]]
%result = "stablehlo.collective_broadcast"(%operand) {
  replica_groups = dense<[[2, 1]]> : tensor<1x2xi64>,
  channel_handle = #stablehlo.channel_handle<handle = 0, type = 0>
} : (tensor1x2xi64>) -> tensor<1x2xi64>
// %result@(0, 0): [[0, 0]]
// %result@(1, 0): [[5, 6]]
// %result@(2, 0): [[5, 6]]
// %result@(3, 0): [[0, 0]]
```

@changm I hope this answers your question!

Reply:
Added

Reply:
> // num_replicas: 4
> // num_partitions: 1

Are we not sure it's not reversed here? My examples before where with 4 partitions and 1 replica.

Reply:
`channel_id = handle = 0` in the above example, so we'd use `cross_replica` instead of `cross_partition`, so we'd want to have multiple replicas here instead.

Reply:
Ok cool, just making sure.

Reply:
For clarification to whoever else may be reading this thread, if you take a look at (C2), if we use `handle = 0` and had `num_replicas = 1`, then the values of `replica_groups` can only be 0. If `handle > 0`, and `num_partitions = 4`, the `replica_groups` can be in range [0, 4).

Reply:
small nit: My understanding of this thread was that an example would be added with handle = 0. That doesn't seem to be the case or was the idea that the handle = 0 case would be in the reference interpreter test cases? Doesn't block either way. Thanks!

Reply:
If I'm understanding correctly, yes the idea is to have handle = 0 (cross_replica) as the example. The reference interpreter would have this test plus other cases (handle > 0, i.e. cross_partition). Let me know if you meant something else.

Reply:
Sounds like this feedback is mostly resolved / non-blocking. Resolving conversation, please unresolve if I am mistaken!